### PR TITLE
IEP-1049: Add a validation for esp-idf path in the install tools dialog

### PR DESCRIPTION
**Pull request by alirana01**
## Description
Added validation in tools installation to check for idf directory

Fixes # ([IEP-1049](https://jira.espressif.com:8443/browse/IEP-1049))

## Type of change

Please delete options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)

## How has this been tested?
Select a directory that is not ESP-IDF the installation dialog should reflect proper error message and valid idf directory should allow installation of tools

**Test Configuration**:
* ESP-IDF Version: any
* OS (Windows,Linux and macOS): all

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [ ] Verified on all platforms - Windows,Linux and macOS

## Summary by CodeRabbit

- **New Features**
  - Enhanced validation logic for IDF tools and paths, providing clearer error messages and improved checks.
  - Introduced a method to dynamically retrieve the IDF version using Git.

- **Refactor**
  - Simplified version validation process and IDF tool path handling.
  - Removed unnecessary method prompting users about spaces in project paths.

- **Bug Fixes**
  - Improved clarity of error messages related to missing IDF Python scripts and improper directory selection.

- **Documentation**
  - Updated user-facing messages for better clarity regarding the installation process and handling of directory paths.

### CODE REVIEW

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java [hunk 1]
```diff
@@ -397,9 +397,10 @@ private void validate()
 			}
 			
 			String idfPyPath = idfPath + File.separator + "tools" + File.separator + "idf.py"; //$NON-NLS-1$ //$NON-NLS-2$
-			if (!new File (idfPyPath).exists())
+			String idfToolsPyPth = idfPath + File.separator + "tools" + File.separator + "idf_tools.py"; //$NON-NLS-1$ //$NON-NLS-2$
+			if (!new File (idfPyPath).exists() && !new File (idfToolsPyPth).exists())
 			{
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java:402]: Use constants for idf.py and idf_tools.py from the IDFConstants file.

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java [hunk 2]
```diff
@@ -397,9 +397,10 @@ private void validate()
 			}
 			
 			String idfPyPath = idfPath + File.separator + "tools" + File.separator + "idf.py"; //$NON-NLS-1$ //$NON-NLS-2$
-			if (!new File (idfPyPath).exists())
+			String idfToolsPyPth = idfPath + File.separator + "tools" + File.separator + "idf_tools.py"; //$NON-NLS-1$ //$NON-NLS-2$
+			if (!new File (idfPyPath).exists() && !new File (idfToolsPyPth).exists())
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java:401]: We need to show an error message even if one of the files does not exist.

```
if (!new File(idfPyPath).exists() || !new File(idfToolsPyPath).exists())
{
    // show error message
}
```

**Review by kolipakakondal — CHANGES_REQUESTED** [review]

**Comment by kolipakakondal** [comment]
@AndriiFilippov please verify once the PR comments are addressed!

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java (hunk 2 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java:401]: @kolipakakondal Thanks for the feedback but we are already setting error message on wizard do you think we should show another popup message to user on an already opened popup wizard?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java (hunk 1 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java:403]: Done

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java (hunk 2 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java]: @kolipakakondal Done

**Review by kolipakakondal — APPROVED** [review]
LGTM @alirana01

**Comment by AndriiFilippov** [comment]
@alirana01 hi ! 

Tested under: 
OS - Windows 10 



select any folder with white space. The white space validation was executed. Then, the absence of the esp-idf folder was not validated, allowing the process to proceed with the incorrect folder.

image

THEN: 
image

**Comment by kolipakakondal** [comment]
> nce of the esp-idf folder was no

@alirana01 Can you please look into this? 

Can we also avoid showing error message popups on the wizard? Instead, we could show error messages in the wizard's message area at the top.?

**Comment by AndriiFilippov** [comment]
@alirana01 hi ! 

Tested under: 
OS - Windows 10 



It now let you to select ESP-IDF with spaces in path like this:
image


BUT downloading validation a bit broken.  
to reproduce: select release 4.4 and folder with spaces.
image

and now switch version to >=5.0 - bug ⛔ 
image


The same works for opposite way:  select >=5.0 - it pass validation and then select v4.4 - no validation  
image

**Comment by AndriiFilippov** [comment]
Probably we should delete this part of info message. It does not make sense. 
image

**Comment by alirana01** [comment]
@AndriiFilippov please verify after the builds and also make sure to test on mac and linix

**Comment by AndriiFilippov** [comment]
@alirana01 hi !

Tested under:
OS - Windows 11 / MacOS

Well, since we have removed `releasev4.4` from Download versions list - it became more easier :D 

I don't like only this behave- when we select just empty folder(with whitespaces) the validation starts for "release <5.0" not the validation for esp-idf folder. 

image


So, my point is: the validation for release <v5.0 should be just in case if the v4.4 selected in a folder with whitespace like here: 
image

But if the empty folder selected, this message does not have any sense: 

image

instead should be: 
image

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java [hunk 3]
```diff
@@ -455,6 +482,95 @@ private void validate()
...
+		{
+			Logger.log("Executing commads: " + commands.toString());
+			Map<String, String> environment = new HashMap<>(System.getenv());
+			IStatus status = processRunner.runInBackground(commands, Path.fromOSString(idfPath), environment);
+			if (status == null)
+			{
+				Logger.log(IDFCorePlugin.getPlugin(),
+						IDFCorePlugin.errorStatus("Unable to get the process status.", null)); //$NON-NLS-1$
+			}
+			String cmdOutput = status.getMessage();
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java:505]: we have check for status == null but we are not using break in that condition, so here is possible NPE

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tools/ToolsJob.java [hunk 4]
```diff
@@ -107,31 +107,40 @@ protected void processExportCmdOutput(final String exportCmdOp)
...
+//			if (matcher.find())
+//			{
+//				idfToolSet.setIdfVersion(matcher.group(1));
+//			}
+//			idfToolSet.getEnvVars().put(IDFEnvironmentVariables.ESP_IDF_VERSION, idfToolSet.getIdfVersion());
+//		}
+//		else
+//		{
+//			idfToolSet.setIdfVersion(idfToolSet.getEnvVars().get(IDFEnvironmentVariables.ESP_IDF_VERSION));	
+//		}
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tools/ToolsJob.java:134]: can we remove the commented code?

**Review by sigmaaa — COMMENTED** [review]
LGTM.  Please take a look at the comments

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java (hunk 3 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/install/IDFDownloadPage.java:506]: Updated

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tools/ToolsJob.java (hunk 4 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tools/ToolsJob.java]: Removed

**Comment by AndriiFilippov** [comment]
@alirana01 hi !

Tested under:
OS - Windows 11 / MacOS arm64


LGTM 👍

**Review by AndriiFilippov — APPROVED** [review]

**Pull request merged**